fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -431,10 +440,85 @@ protected void handleResultsResponse(
             );
             return Observable.error(responseException);
         }
+
+        if (jobsApiRequest.getPaginationParameters().isPresent()) {
+            Pagination<Result> pages = new AllPagesPagination<>(
+                    preResponse.getResultSet(),
+                    jobsApiRequest.getPaginationParameters().get()
+            );
+            responseContext = paginateResponseContext(responseContext, jobsApiRequest, pages);
+            PreResponse paginatedPreResponse = new PreResponse(
+                    new ResultSet(pages.getPageOfData(), preResponse.getResultSet().getSchema()),
+                    responseContext
+            );
+            return Observable.just(paginatedPreResponse);
+        }
         return Observable.just(preResponse);
     }
 
     /**
+     * Add pagination details to ResponseContext.
+     *
+     * @param responseContext  ResponseContext object contains all the meta info of the resultSet
+     * @param jobsApiRequest  JobsApiRequest object with all the associated info with it
+     * @param pages  Paginated resultSet
+     *
+     * @return Updated ResponseContext contains pagination info
+     */
+    protected ResponseContext paginateResponseContext(
+            ResponseContext responseContext,
+            ApiRequest jobsApiRequest,
+            Pagination<Result> pages
+    ) {
+        LinkedHashMap<String, URI> bodyLinks = new LinkedHashMap<>();
+        Arrays.stream(PaginationLink.values())
+                .forEachOrdered(link -> addLink(
+                                link,
+                                pages,
+                                jobsApiRequest.getUriInfo().getRequestUriBuilder(),
+                                bodyLinks)
+                );
```

Comment:
👍 (though it would be probably a good idea to comment a bit on each step as to what's happening, since the `Utils.withRight` nonsense makes it more obscure than it should be (thanks Java...).

Alternatively, another approach would be to break those operations out into their own methods that would encapsulate the operation in a nice, descriptive name.